feature: Supporting core+package templates stored on NuGet

pavolum opened the pull request:
## Description

Core of this PR:

- Adds a new type of template to Composer called Conversational Core
    - Conversational Core is stored as a public nuget package and is made up of a core bot with dependent packages
    - Conversational Core is a dotnet template that starts off with an ejected runtime that is packaged with the template
- Adds logic to Composer to support this new type of template
    - If the users selects this template in the creation flow and selects 'create' then Composer 
        - installs that dotnet template stored on NuGet to the users machine if it does not already exist (dotnet new -u <nugetPackageName>)
        - Creates an instance of that template and stores the resulting bot in memory.
        - Runs the dialog merge tool to grab any dependent packages that the template declares and merge them with the bot in memory (for conversational core there are 4 lg packages that are merged at this step)
        - User is outputted to editing view where they see the core bot and the 4 lg assets ready to be edited or tested

Side things that the PR does:
- Consolidates redundant types BotProject and BotTemplate
- Allows for relative paths in settings.runtime.path which is required to support a template that comes ejected on initialization
- Adds a public feature flag that toggles all externally stored templates from being visible in the creation flow 
- Updates feature flag defaults
- Simplifies feature flag init logic 

## Task Item

fixes #4371 
fixes #4567

hatpick commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -52,13 +57,52 @@ export class AssetManager {
+    if (isHostedTemplate) {
+      // create empty temp directory on server for holding externally hosted template src
+      templateSrcPath = path.resolve(__dirname, '../../../temp');
+      fs.mkdir(templateSrcPath, (err) => {
```
should await here? (use promiseified version from fs-extra perhaps)

hatpick commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -109,19 +153,21 @@ export class AssetManager {
+      const location = Path.join(boilerplate.path, 'scripts', 'package.json');
+      try {
+        if (fs.existsSync(location)) {
+          const raw = fs.readFileSync(location, 'utf8');
```
maybe avoid using sync version in case it's a big file to read?

srinaath replied:
@hatpick Its a very small file basically  just a blank template. In addition, we read it once and inside the getter above it always picks the cached value after the first time

srinaath replied:
I do think I can change it to async for reading it the first time. But its not a P0 IMO.

hatpick replied:
Ok, my thinking is that since it is on the server side and potentially many users will hit the sync read, it may cause perf issues down the road.

hatpick commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -131,15 +177,17 @@ export class AssetManager {
-        const json = JSON.parse(raw);
-        return json;
+    if (boilerplate.path) {
+      const location = Path.join(boilerplate.path, `${boilerplate.id}.botproj`);
```
instead of hard coding the extension, use FileExtensions

hatpick commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -131,15 +177,17 @@ export class AssetManager {
+      const location = Path.join(boilerplate.path, `${boilerplate.id}.botproj`);
+      try {
+        if (fs.existsSync(location)) {
+          const raw = fs.readFileSync(location, 'utf8');
```
same here

hatpick commented on Composer/packages/types/src/runtime.ts:
```diff
@@ -37,11 +37,15 @@ export type BotTemplate = {
   tags?: string[];
   /* list of supported runtime versions */
   support?: string[];
+  packageName?: string;
```
I'm wondering if you can do:
```
package?: {
  name: string;
  source: string;
  version: string;
}
```

This way when there is a package, there must be a name, version and source.

hatpick commented on Composer/packages/client/src/recoilModel/selectors/projectTemplates.ts:
```diff
@@ -18,6 +18,16 @@ export const filteredTemplatesSelector = selector({
+    if (!featureFlags?.REMOTE_TEMPLATE_CREATION_EXPERIENCE?.enabled) {
+      let index = filteredTemplates.length - 1;
+
+      while (index >= 0) {
```
can you use .filter

pavolum replied:
Sure thing

hatpick commented on Composer/packages/client/src/recoilModel/selectors/projectTemplates.ts:
```diff
@@ -18,6 +18,16 @@ export const filteredTemplatesSelector = selector({
         filteredTemplates.splice(vaTemplateIndex, 1);
       }
     }
+    if (!featureFlags?.REMOTE_TEMPLATE_CREATION_EXPERIENCE?.enabled) {
```
can featureFlags not have a value?

hatpick replied:
can featureFlags?.REMOTE_TEMPLATE_CREATION_EXPERIENCE not have a value?

hatpick commented on Composer/packages/server/src/controllers/project.ts:
```diff
@@ -54,6 +55,33 @@ async function createProject(req: Request, res: Response) {
     await AssetService.manager.copyBoilerplate(currentProject.dataDir, currentProject.fileStorage);
 
     if (currentProject !== undefined) {
+      if (currentProject?.settings?.runtime.customRuntime === true) {
```
you just made sure currentProject is defined, .? is unnecessary

hatpick replied:
can a project not have settings?

pavolum replied:
Removed the null checks, first is not needed and all bot projects are loaded with default settings if their template does not prepopulate

hatpick commented on Composer/packages/server/src/models/bot/botProject.ts:
```diff
@@ -199,6 +199,15 @@ export class BotProject implements IBotProject {
     const settings = await this.settingManager.get(obfuscate);
 
+    // Resolve relative path for custom runtime if the path is relative
+    if (settings?.runtime?.customRuntime && settings.runtime.path && !Path.isAbsolute(settings.runtime.path)) {
```
first condition is worried that the settings is undefined, second and third conditions are not?

hatpick commented on Composer/packages/types/src/runtime.ts:
```diff
@@ -37,11 +37,15 @@ export type BotTemplate = {
   description: string;
   /* absolute path */
-  path: string;
+  path?: string;
```
is it optional?

pavolum commented:
Placeholder while CI builds are down:

Test:Coverage ran on this PR through the CI earlier and showed these two tests breaking, the fixes for these tests are in but the CI build is down and wont run test: coverage. I have fixed these tests and tested the fixes locally (output below).

Fixed test output
```
 PASS   server  ../../services/__tests__/project.test.ts (12.771 s)
 PASS   server  ./project.test.ts (16.73 s)
 PASS   server  ../../models/bot/__tests__/botProject.test.ts (21.783 s)

Test Suites: 3 passed, 3 total
Tests:       50 passed, 50 total
Snapshots:   0 total
Time:        23.326 s, estimated 28 s
Ran all test suites matching /project.test.ts/i.
```
```
 PASS   client  ../DialogWrapper/index.test.tsx (8.085 s)
 PASS   client  DefineConversation/index.test.tsx (15.57 s)
 PASS   client  CreateOptions/index.test.tsx (15.702 s)
 PASS   client  ./index.test.tsx (17.988 s)

Test Suites: 4 passed, 4 total
Tests:       7 passed, 7 total
Snapshots:   0 total
Time:        20.113 s, estimated 41 s
Ran all test suites matching /index.test.tsx/i.
```

test:coverage output
```
Summary of all failing tests
FAIL packages/server/src/controllers/__tests__/project.test.ts
  ● create a Empty Bot project › should create a new project

    expect(jest.fn()).toHaveBeenCalledWith(...expected)

    Expected: 200
    Received: 404

    Number of calls: 1

      177 |     } as Request;
      178 |     await ProjectController.createProject(mockReq, mockRes);
    > 179 |     expect(mockRes.status).toHaveBeenCalledWith(200);
          |                            ^
      180 |   });
      181 | });
      182 | //current opened bot is the newBot

      at Object.toHaveBeenCalledWith (src/controllers/__tests__/project.test.ts:179:28)

FAIL packages/client/__tests__/components/CreationFlow/index.test.tsx
  ● <CreationFlow/> › should render the component

    TypeError: Cannot read property 'enabled' of undefined

      20 |       }
      21 |     }
    > 22 |     if (!featureFlags.REMOTE_TEMPLATE_CREATION_EXPERIENCE.enabled) {
         |                                                           ^
      23 |       filteredTemplates = filteredTemplates.filter((template: BotTemplate) => {
      24 |         if (template.path) {
      25 |           return template;

      at enabled (src/recoilModel/selectors/projectTemplates.ts:22:59)
      at evaluateSelectorFunction (../../node_modules/recoil/cjs/recoil.js:2981:22)
      at getFromCacheOrEvaluate (../../node_modules/recoil/cjs/recoil.js:2947:44)
      at ../../node_modules/recoil/cjs/recoil.js:3073:47
      at detectCircularDependencies (../../node_modules/recoil/cjs/recoil.js:3039:14)
      at Object.myGet [as get] (../../node_modules/recoil/cjs/recoil.js:3073:14)
      at getNodeLoadable (../../node_modules/recoil/cjs/recoil.js:306:25)
      at getRecoilValueAsLoadable (../../node_modules/recoil/cjs/recoil.js:714:37)
      at useRecoilValueLoadable_LEGACY (../../node_modules/recoil/cjs/recoil.js:2056:10)
      at useRecoilValueLoadable (../../node_modules/recoil/cjs/recoil.js:2073:12)
      at useRecoilValue (../../node_modules/recoil/cjs/recoil.js:2090:20)
      at useRecoilValue (src/components/CreationFlow/CreationFlow.tsx:51:28)
      at renderWithHooks (../../node_modules/react-dom/cjs/react-dom.development.js:14803:18)
      at mountIndeterminateComponent (../../node_modules/react-dom/cjs/react-dom.development.js:17482:13)
      at beginWork (../../node_modules/react-dom/cjs/react-dom.development.js:18596:16)
      at beginWork$1 (../../node_modules/react-dom/cjs/react-dom.development.js:23179:14)
      at performUnitOfWork (../../node_modules/react-dom/cjs/react-dom.development.js:22157:12)
      at workLoopSync (../../node_modules/react-dom/cjs/react-dom.development.js:22130:22)
      at performSyncWorkOnRoot (../../node_modules/react-dom/cjs/react-dom.development.js:21756:9)
      at scheduleUpdateOnFiber (../../node_modules/react-dom/cjs/react-dom.development.js:21188:7)
      at updateContainer (../../node_modules/react-dom/cjs/react-dom.development.js:24373:3)
      at ../../node_modules/react-dom/cjs/react-dom.development.js:24758:7
      at unbatchedUpdates (../../node_modules/react-dom/cjs/react-dom.development.js:21903:12)
      at legacyRenderSubtreeIntoContainer (../../node_modules/react-dom/cjs/react-dom.development.js:24757:5)
      at Object.render (../../node_modules/react-dom/cjs/react-dom.development.js:24840:10)
      at ../../node_modules/@testing-library/react/dist/pure.js:86:25
      at batchedUpdates$1 (../../node_modules/react-dom/cjs/react-dom.development.js:21856:12)
      at act (../../node_modules/react-dom/cjs/react-dom-test-utils.development.js:929:14)
      at render (../../node_modules/@testing-library/react/dist/pure.js:82:26)
      at renderWithRouter (__tests__/components/CreationFlow/index.test.tsx:48:10)
      at renderWithRouter (__tests__/components/CreationFlow/index.test.tsx:61:9)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:62:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:288:22)
      at Generator.prototype.<computed> [as next] (node_modules/regenerator-runtime/runtime.js:114:21)
      at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at Object.<anonymous> (node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12)


Test Suites: 2 failed, 224 passed, 226 total
Tests:       2 failed, 1 todo, 901 passed, 904 total
Snapshots:   5 passed, 5 total
Time:        439.651 s
Ran all test suites in 13 projects.
Force exiting Jest: Have you considered using `--detectOpenHandles` to detect async operations that kept running after all tests finished?
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

pavolum commented:
@hatpick @cwhitten @lzc850612 Unit test build passed, but others did not due to myget being down. This code has been reviewed and does not break tests. 

@hatpick approve when ready and then @cwhitten can squash and merge with his mighty admin privilege's.
